### PR TITLE
ETCD-58: Add enhancement for automated defragmentation

### PR DESCRIPTION
**Pull request by hexfusion**
This enhancement is the documentation for the defragmentation controller which shipped in 4.9.

### CODE REVIEW

File: enhancements/etcd/automated-defrag.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+### Implementation Details/Notes/Constraints
+
+## etcd Storage Backend
+etcd's storage backend is powered by `bbolt` a B+Tree memory mapped key/value store with MVCC support. While MVCC is
+necessary to support fully concurrent transactions and the heart of the `Watch` functionality.
+MVCC poses a performance challenge for `bbolt` because the db can store many versions of the keys value. These
+versions can build up indefinitely and result in memory and disk performance issues. The solution is to regularly
+compact the key space.
+
+compaction is essentially a delete action in `bbolt`, where the prevision versions of a keys value are pruned.
```
Comment (lilic) [inline enhancements/etcd/automated-defrag.md:73]: Nit:

```suggestion
Compaction is essentially a delete action in `bbolt`, where the prevision versions of a keys value are pruned.
```

File: enhancements/etcd/automated-defrag.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    // dbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+    DbSizeInUse int64 `protobuf:"varint,9,opt,name=dbSizeInUse,proto3" json:"dbSizeInUse,omitempty"`
+}
+```
+
+From these two fields we can conclude the total fragmented space on that etcd members backend store. If the
+percentage of bytes of the total keyspace are over a defined threshold percentage. The controller will 
+utilize maintenance API again and issue a `DefragmentRequest` RPC.
+
+Defragmentation is I/O blocking thus important to only defrag a single etcd at a time. Additionally, to reduce
```
Comment (lilic) [inline enhancements/etcd/automated-defrag.md:100]: How is this going to be achieved, to restrain defrag to a single etcd at a time? Could we detail that a bit, sorry if I missed it somewhere?

File: enhancements/etcd/automated-defrag.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+defragment the store is directly related to the amount of free pages that exists. For this reason we feel it
+makes sense to continuously observe the clusters state and defrag frequently along the cadence of 
+compaction resulting in shorter intervals of disruption by keeping the number of pages limited and the size of the 
+database smaller.
+
+The criteria for defragmentation is defined as:
+
+- Cluster must be observed as `HighlyAvailableTopologyMode`.
+- The etcd cluster must report all members as healthy.
+- `minDefragBytes` this is a minimum database size in bytes before defragmentation should be considered.
```
Comment (lilic) [inline enhancements/etcd/automated-defrag.md:111]: What is this value?

File: enhancements/etcd/automated-defrag.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+TODO show performance graphs.
+
+### Risks and Mitigations
+
+- SNO: As defragmentation is I/O blocking it is unsafe for non HA clusters. Mitigation to this risk includes a check in 
+  the controller for control-plane topology and will exit 0 if non HA.
+  
+- datafile corruption: Recently a bug was fixed in upstream etcd[1] which could lead to corruption of the key space 
+  as a result of defragmentation if it happened while etcd was being terminated. Mitigation to this risk includes 
+  heavy exercise in CI and telemetry will alert `etcdMemberDown` if one of the etcd instances goes down as the
```
Comment (lilic) [inline enhancements/etcd/automated-defrag.md:124]: etcdMemberDown fires after 1 minute, is this short or long enough for this?

**Review by lilic — COMMENTED** [review]
Great work and really detailed proposal! Few questions from my side only.

File: enhancements/etcd/automated-defrag.md (hunk 3 above)
Comment (hexfusion) [reply enhancements/etcd/automated-defrag.md:111]: I am thinking 100MB to reduce early defrag during install. Its possible this is not needed but I think 100MB would be reasonable.

Comment (lilic) [reply enhancements/etcd/automated-defrag.md:111]: Sounds good, now I see the need to send this data back to telemetry, thanks!

File: enhancements/etcd/automated-defrag.md (hunk 4 above)
Comment (hexfusion) [reply enhancements/etcd/automated-defrag.md:124]: the timeout for the call is 45s technically we could see a wait that long but we would miss 1m.

File: enhancements/etcd/automated-defrag.md (hunk 2 above)
Comment (hexfusion) [reply enhancements/etcd/automated-defrag.md:100]: will add in more detail

**Review by lilic — APPROVED** [review]
/lgtm

**Review by lilic — APPROVED** [review]
/lgtm

Thanks!

**Comment by hexfusion** [comment]
/remove-lifecycle rotten

This needs to get merged

**Comment by dhellmann** [comment]
@hexfusion is it in a state where it's ready to go now? I see @deads2k is listed as an approver, does anyone else need to look at it?

**Comment by hexfusion** [comment]
@dhellmann This is more or less documentation as the feature will be shipped in 4.9.

cc @mfojtik

**Comment by sdodson** [comment]
/approved
Assuming since this shipped in 4.9 we should land this, if there are material differences between final impelmentation and what's defined in this enhancement please find some time to update the enhancement contents.

**Comment by sdodson** [comment]
The linter appears to only check modified files so I'm going tomerge this, whomever touches this file next will have to clean up the linting problem.

**Pull request merged**